src/hotspot/share/runtime/crac.cpp

```diff
@@ -422,6 +428,8 @@ Handle crac::checkpoint(jarray fd_arr, jobjectArray obj_arr, bool dry_run, jlong
   Universe::heap()->set_cleanup_unused(false);
   Universe::heap()->finish_collection();
 
+  AsyncLogWriter::instance()->flush();
```

Comment:
As a non-java thread, AsyncLogWriter does not participate in the safepoint protocol. What happens if another non-java thread enqueues a message to be written while the output is closed?

Reply:
OK, I have to improve it, thanks.

Reply:
I hope the stop() and resume() do fix it. But there is no testcase for that, do you want it?

Reply:
I guess that a testcase for the synchronization would be quite difficult to do, as you'd be trying to simulate a race.

However, I am not convinced the `stop()` and `resume()` work correctly: the log writer does not need to own the lock while writing messages to the (potentially nulled) outputs; you flush the buffer and then acquire the lock, effectively blocking all log producers from enqueuing further messages. However if someone enqueues a message right after flush, the log writer has a chance to dequeue it (actually swap queues) and even if you acquire lock then, continue writing these messages to the closed output.

Reply:
Race reproducers are also a part of testsuites - [ptrace-testsuite](https://sourceware.org/systemtap/wiki/utrace/tests) limits their run by its `TESTTIME` parameter. I sure did not try how easily would be to reproduce this race and I am sure happy not to write such a testcase.